analyzers/src/SonarAnalyzer.CFG/ShimLayer/AnalysisContext/SymbolStartAnalysisContext.cs

```diff
@@ -0,0 +1,169 @@
+        registerSyntaxNodeActionCS = CreateRegistrationMethodWithAdditionalParameter<SyntaxNodeAnalysisContext, ImmutableArray<CS.SyntaxKind>>(
+            symbolStartAnalysisContextType, nameof(RegisterSyntaxNodeAction), typeof(CS.SyntaxKind));
+    }
+
+    public SymbolStartAnalysisContext(object roslynSymbolStartAnalysisContext)
+    {
+        RoslynSymbolStartAnalysisContext = roslynSymbolStartAnalysisContext;
+    }
+    public object RoslynSymbolStartAnalysisContext { get; }
+    public CancellationToken CancellationToken => cancellationTokenAccessor(RoslynSymbolStartAnalysisContext);
+    public Compilation Compilation => compilationAccessor(RoslynSymbolStartAnalysisContext);
+    public AnalyzerOptions Options => optionsAccessor(RoslynSymbolStartAnalysisContext);
+    public ISymbol Symbol => symbolAccessor(RoslynSymbolStartAnalysisContext);
+
+    public void RegisterCodeBlockAction(Action<CodeBlockAnalysisContext> action) =>
+        registerCodeBlockAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterCodeBlockStartAction<TLanguageKindEnum>(Action<CodeBlockStartAnalysisContext<TLanguageKindEnum>> action) where TLanguageKindEnum : struct
+    {
+        var languageKindType = typeof(TLanguageKindEnum);
+        if (languageKindType == typeof(CS.SyntaxKind))
+        {
+            var casted = (Action<CodeBlockStartAnalysisContext<CS.SyntaxKind>>)action;
+            registerCodeBlockStartActionCS(RoslynSymbolStartAnalysisContext, casted);
+        }
+        else if (languageKindType.FullName == "Microsoft.CodeAnalysis.VisualBasic.SyntaxKind")
+        {
+            throw new NotImplementedException("Add a reference to the Microsoft.CodeAnalysis.VisualBasic.Workspaces package.");
+        }
+        else
+        {
+            throw new ArgumentException("Invalid type parameter.", nameof(TLanguageKindEnum));
+        }
+    }
+
+    public void RegisterOperationAction(Action<OperationAnalysisContext> action, ImmutableArray<OperationKind> operationKinds) =>
+        registerOperationAction(RoslynSymbolStartAnalysisContext, action, operationKinds);
+
+    public void RegisterOperationBlockAction(Action<OperationBlockAnalysisContext> action) =>
+        registerOperationBlockAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterOperationBlockStartAction(Action<OperationBlockStartAnalysisContext> action) =>
+        registerOperationBlockStartAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterSymbolEndAction(Action<SymbolAnalysisContext> action) =>
+        registerSymbolEndAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterSyntaxNodeAction<TLanguageKindEnum>(Action<SyntaxNodeAnalysisContext> action, params TLanguageKindEnum[] syntaxKinds) where TLanguageKindEnum : struct
+    {
+        var languageKindType = typeof(TLanguageKindEnum);
```

Comment:
The following logic of comparing `typeof(TLanguageKindEnum)` against `CS.SyntaxKind`, then its name again an hardcoded string `Microsoft.CodeAnalysis.VisualBasic.SyntaxKind`, and returning a `ArgumentException` otherwise is duplicated.
I would factorize into a `EnsureCSharpLanguageKind` method. That should be easy because `languageKindType` is used only for the checks and nowhere else in either of the two methods.

Reply:
The downside is, that we actually should implement VB right away and if we do so, the methods need to separated as they are at the moment: There is nothing common to extract (at least in C#).